File: iterator.go
package bitcask

import (
	"bytes"

	"github.com/yifaaan/bitcask/index"
)

// 迭代器
type Iterator struct {
	// 索引迭代器
	indexIter index.Iterator
	db        *DB
	options   IteratorOptions
}

func (db *DB) NewIterator(opts IteratorOptions) *Iterator {
	indexIter := db.index.Iterator(opts.Reverse)
	return &Iterator{indexIter: indexIter, db: db, options: opts}
}

// 重新回到迭代器起点
func (it *Iterator) Rewind() {
	it.indexIter.Rewind()
	it.skipToNext()
}

// 根据key找到第一个大于（或小于）等于的目标
func (it *Iterator) Seek(key []byte) {
	it.indexIter.Seek(key)
	it.skipToNext()
}

// 下一个key
func (it *Iterator) Next() {
	it.indexIter.Next()
	it.skipToNext()
}

// 是否已经遍历完
func (it *Iterator) Valid() bool {
	return it.indexIter.Valid()
}

// 当前位置的key
func (it *Iterator) Key() []byte {
	return it.indexIter.Key()
}

// 当前位置的数据
func (it *Iterator) Value() ([]byte, error) {
	logRecordPos := it.indexIter.Value()
	it.db.mu.RLock()
	defer it.db.mu.RUnlock()
	return it.db.getValueByPosition(logRecordPos)
}

// 关闭
func (it *Iterator) Close() {
	it.indexIter.Close()
}

// 找到以prefix为前缀的项
func (it *Iterator) skipToNext() {
	prefixLen := len(it.options.Prefix)
	// 前缀为空，不用跳过
	if prefixLen == 0 {
		return
	}

	for ; it.indexIter.Valid(); it.indexIter.Next() {
		key := it.Key()
		if prefixLen <= len(key) && bytes.Equal(it.options.Prefix, key[:prefixLen]) {
			return
		}
	}
}

File: src/iterator.cpp
#include "iterator.h"

#include <absl/status/statusor.h>

#include <shared_mutex>

#include "db.h"

namespace bitcask
{
    Iterator::Iterator(DB* db, std::unique_ptr<IndexIterator> index_iter, IteratorOptions options) : db_(db), index_iter_(std::move(index_iter)), options_(std::move(options))
    {
    }

    void Iterator::Rewind()
    {
        index_iter_->Rewind();
        SkipToNext();
    }

    void Iterator::Seek(absl::string_view key)
    {
        index_iter_->Seek(key);
        SkipToNext();
    }

    void Iterator::Next()
    {
        index_iter_->Next();
        SkipToNext();
    }

    [[nodiscard]] bool Iterator::Valid() const
    {
        return index_iter_->Valid();
    }

    [[nodiscard]] absl::string_view Iterator::Key() const
    {
        return index_iter_->Key();
    }

    [[nodiscard]] std::optional<std::string> Iterator::Value() const
    {
        auto pos = index_iter_->Value();
        absl::ReaderMutexLock lock(db_->mutex_);
        if (auto res = db_->GetValueByPosition(pos); res.ok())
        {
            return res.value();
        }
        return std::nullopt;
    }

    
    void Iterator::SkipToNext()
    {
        if (options_.prefix.empty())
        {
            return;
        }
       for (; index_iter_->Valid(); index_iter_->Next())
       {
            const auto key = index_iter_->Key();
            if (key.starts_with(options_.prefix))
            {
                break;
            }
       }

    }
}
